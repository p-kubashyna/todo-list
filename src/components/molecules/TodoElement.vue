<script setup>
import CheckboxInput from '@/components/atoms/CheckboxInput.vue'
import ButtonInput from '@/components/atoms/ButtonInput.vue'
import { useStore } from 'vuex'

const store = useStore()
const props = defineProps({
  todoItem: {
    type: Object,
    required: true
  }
})

function removeTodo (itemId) {
  store.commit('removeTodo', itemId)
}

function changeTodoState (itemId) {
  store.commit('changeState', itemId)
}
</script>

<template>
  <div class="item">
    <CheckboxInput :checked="props.todoItem.isCompleted" @change="changeTodoState(props.todoItem.id)"/>
    <span :class="{ 'completed-task' : props.todoItem.isCompleted === true }">{{ props.todoItem.name }}</span>
    <ButtonInput class="button-default" label="X" @click="removeTodo(props.todoItem.id)"/>
  </div>
</template>

<style scoped>
.item {
  height: 30px;
  background-color: white;
  border-width: 0 0 1px 0;
  border-color: gray;
  border-style: solid;
  padding: 7px;
  display: flex;
  align-items: center;
}

.button-default {
  width: 25px;
  background-color: white;
  border-width: 0;
}

.completed-task {
  text-decoration: line-through;
  color: grey;
}
</style>
