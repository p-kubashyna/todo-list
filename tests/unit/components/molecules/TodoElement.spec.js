import { describe, expect, it } from "@jest/globals";
import { mount } from "@vue/test-utils";
import TodoElement from "@/components/molecules/TodoElement.vue";
import { createStore } from "vuex";

describe("Given the TodoElement component", () => {
  const defaultTodo = {
    todoItem: { name: "todo", isCompleted: false, id: "id" },
  };

  it.each([
    [{ todoItem: { name: "todo", isCompleted: false, id: "id" } }],
    [{ todoItem: { name: "todo", isCompleted: true, id: "id" } }],
  ])("should render", (propsData) => {
    const store = createStore({});
    const wrapper = mount(TodoElement, {
      props: propsData,
      global: {
        plugins: [store],
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should change todo state on checkbox change", async () => {
    const store = createStore({});
    store.commit = jest.fn();
    const wrapper = mount(TodoElement, {
      props: defaultTodo,
      global: {
        plugins: [store],
      },
    });
    await wrapper.find("input[type='checkbox']").trigger("change");
    await wrapper.vm.$nextTick();

    expect(store.commit).toHaveBeenCalledTimes(1);
    expect(store.commit).toHaveBeenCalledWith(
      "changeState",
      defaultTodo.todoItem.id
    );
  });

  it("should delete todo on button click", async () => {
    const store = createStore({});
    store.commit = jest.fn();
    const wrapper = mount(TodoElement, {
      props: defaultTodo,
      global: {
        plugins: [store],
      },
    });
    await wrapper.find("button").trigger("click");
    await wrapper.vm.$nextTick();

    expect(store.commit).toHaveBeenCalledTimes(1);
    expect(store.commit).toHaveBeenCalledWith(
      "removeTodo",
      defaultTodo.todoItem.id
    );
  });
});
