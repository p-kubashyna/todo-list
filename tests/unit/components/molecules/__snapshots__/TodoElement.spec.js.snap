// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the TodoElement component should render 1`] = `
<div
  class="item"
>
  <input
    class="checkbox-round"
    type="checkbox"
  />
  <span
    class="task"
  >
    todo
  </span>
  <button
    class="delete-button"
  >
    <img
      alt="delete icon"
      class="delete-button-image"
      src=""
    />
  </button>
</div>
`;

exports[`Given the TodoElement component should render 2`] = `
<div
  class="item"
>
  <input
    checked=""
    class="checkbox-round"
    type="checkbox"
  />
  <span
    class="task completed-task"
  >
    todo
  </span>
  <button
    class="delete-button"
  >
    <img
      alt="delete icon"
      class="delete-button-image"
      src=""
    />
  </button>
</div>
`;
